### ISSUE TITLE
[INTERNAL] List of 3.0 enhancements and fixes

### ISSUE THREAD

**Issue by pbatard** [pbatard/rufus#868]
__This topic is flagged as INTERNAL and is therefore NOT open for comments.__

__Also, as you can see, there is an awful lot of work still needed for this release compared to the previous ones, which means that the release date of Rufus 3.0 is HIGHLY flexible and likely to be delayed further.__

General
----------
- [x] Drop XP __and Vista__ support
- [x] Add retries for access denied
- [x] ~~Add tooltip with file info. to the status bar (see #853)~~ _We'll extend the dropdown to display the full name &mdash; that should be enough_
- [x] ~~According to [this](https://en.wikipedia.org/wiki/Design_of_the_FAT_file_system#Directory_table), which may also apply to volumes, we may be too restrictive in our label sanitization. Especially © (probably in CP850) seems to be allowed as part of a volume label.~~ _Nah, we get those through `autorun.inf` anyway, and people need to know better than rely on __obsolete__ codepages_.
- [x] Invert <kbd>Alt</kbd>-<kbd>D</kbd> and <kbd>Alt</kbd>-<kbd>X</kbd> cheat modes
- [x] Update the FAQ on release for  <kbd>Alt</kbd>-<kbd>D</kbd>, <kbd>Alt</kbd>-<kbd>X</kbd>, <kbd>Alt</kbd>-<kbd>V</kbd> and <kbd>Alt</kbd>-<kbd>Q</kbd>
- [x] We should do something about delays and error messages incurred by non existing `A:` and `B:` drives.
- [x] ~~We really seem to have a problem with reformatting the drive in Windows 10 after a GPT image with multiple partitions has been written (e.g. `FreeNAS-x86.img` then FreeDOS). Cycling the port seems to help, but there has to be a better way to fall back on our feet...~~ &rarr; _That'll be 3.1 material_
- [x] Double <kbd>Alt</kbd>-<kbd>C</kbd> will shut down Rufus? WTF?
- [x] Switch to SSL and use `rufus.ie` as our base URL
- [x] __Test__ SSL downloads, including update downloads
- [x] Add ARM64 support to our UEFI:NTFS image (would be nice if we could have F2FS support integrated into GRUB before we do that) _Done!_
- [x] Need to add new detection for `efiboot.img` as per [this](https://www.reddit.com/r/linuxmasterrace/comments/7hqfpj/void_linux_iso_flagged_as_nonuefi_by_rufus/dqu09sk/).
- [x] Get a new code signing certificate _(__COMODO STANDOFF DAY 22:__ Sheesh! How many times do I have to insist with Comodo until they accept that it's not because you aren't in Duns and Bradstreet that your company is invalid!!! And people give me bad press about improbable security issues because they believe it's all too easy to get a non EV code signing certificate under any name you want. Try to have a chat with docs@comodo.com, then, and see how long it takes before you just want to bash your head against the wall, from having to repeat the same thing over and over again!... Oh and now that they do have the phone number, they can't even seem to sort out the international prefix?!? How INCOMPETENT can these people be?)_  
__AT LONG FRIGGING LAST!!!! HEED MY ADVICE: NEVER PURCHASE A CERTIFICATE FROM COMODO!__
- [x] Remove the blurb about GPT/UEFI from the home page
- [x] Update platform support to start at Windows 7
- [x] I've seen a `The downloaded executable is missing a digital signature.` message while testing a 2.18 &rarr; 3.0_BETA update (but only once). Should we be concerned about this? &rarr; _Moved to #1130_
- [x] Confirm what happens when Vista/XP users attempt to check for 3.0 BETA. Is the `platform_min = 6.1` filter enough or do we need to multiply `.ver` files so that Vista/XP users only get 2.18 &rarr; _`platform_min` seems to do the trick - Good job me!_

[UI Redesign](https://github.com/pbatard/rufus/tree/ui_redesign)
----------
__Pre Translator Alpha:__
- [x] Replace disc icon with SELECT
- [x] Center horizontal lines for separators
- [x] Use globals for bt, fs and stuff
- [x] Add Image Settings section
- [x] Remove `#` button on status bar
- [x] Fix log window not being the same height as the main window
- [x] Keep Image settings if switching image
- [x] Keep a list of recent images
- [x] Revert Image Settings section and add the hash button in boot type dropdown
- [x] ~~Add tooltip with full path for each image~~ _Nah, dropdown resize should be good enough. Will do if requested_
- [x] Add tooltip for hash
- [x] Split Advanced Options between Device and Format
- [x] Add new registry keys for Device and Format advanced options
- [x] Add regular up/down toolbars to display advanced options in each section
- [x] Remove image name from boot dropdown if new selection failed _Or rather, keep previously selected image_
- [x] Resize image dropdown according to image length when changing language
- [x] File system population is broken!
- [x] Hash button is now broken!
- [x] ~~Automatic vertical spacing of elements according to height~~
- [x] Move _Create Extended Label and Icon_ to Advanced Section
- [x] ~~Move _Quick Format_ and _List USB Hard Drives_ to same line?~~
- [x] What the heck happened to our _Save_ button?
- [x] Tie action to Save button and remove <kbd>Alt</kbd>-<kbd>V</kbd> cheat mode
- __Main Dialog__
  - [x] Dynamically resize Device dropdown according to whether Save button is displayed or not
  - [x] Dynamically compute the minimum size for the 3 master buttons
  - [x] Dynamically compute the minimum size for the half dropdowns
  - [x] Dynamically compute the minimum size for full width dropdowns
  - [x] I don't think we have a choice but to compute the actual with of an empty dropdown, including the `⌵` symbol, to figure out proper sizing with text
  - [x] Dynamically compute the minimum size for boot type dropdown (see French for empty image)
  - [x] ~~Use spacing from Windows standard Save dialog as guideline for our own~~ _Not sure we want to do that. or at least, not for the main dialog_
  - [x] ~~Should we have the half dropdown separators be 3x the size of small_sep, as well as small_sep be the size of the `#` button? The goal here is to have the space between the boot selection dropdown and the select button, which includes the `#` button, be the same size as the half dropdown sep.~~ _Nah, just use the sep for Save button as our base._
  - [x] Update half-width according to 2x min button size + small_sep
  - [x] Align the _START_ & _CLOSE/CANCEL_ buttons to half-width dropdowns
  - [x] Dynamically space master buttons
  - [x] Dynamically space half dropdowns
  - [x] Dynamically set dialog width
  - [x] Properly resize the device dropdown on init
  - [x] Sort the status bar on dialog resize
  - [x] `?` is not being moved on dialog resize
- __About Dialog__
  - [x] Dynamically resize License ~~and OK~~ buttons
- __License Dialog__
  - [x] ~~Dynamically resize Close button~~ _Not needed_
- __Log Dialog__
  - [x] Remove the unneeded 'Log' from the translations that use it
  - [x] ~~Dynamically resize buttons~~ _Looks like we can get away without a resize. Lazy way out, here I come!_
- __Notification Dialog__
  - [x] Dynamically resize buttons
- __Update Settings Dialog__
  - [x] Dynamically resize dropdowns
  - [x] Dynamically resize buttons
  - [x] Dynamically resize groups
  - [x] Dynamically resize dialog width
- __New Version Dialog__
  - [x] Do we have anything to do there?
- [x] Remove handling of `m` and `s` loc commands
- [x] Use _Show/Hide_ action words on advanced toolbars
- [x] Create tooltip for Save button
- [x] Overlay of status text on progress bar
- [x] Create a flat toolbar buttons for About/Settings/Language/Log
- [x] Map _Settings_ toolbar button to the Update Settings dialog
- [x] Fix multi-toolbar tooltips
- [x] Get the section horizontal lines to match `full_width` alignment
- [x] Fix the width of the _Show Advanced Options_ toolbars using `TB_GETIDEALSIZE`
- [x] Remove the LOG and ABOUT buttons
- [x] ~~Can we use ellipses for the ISOs displayed in the dropdown?~~ _Not easily. Will have to do without for now and use custom draw if users complain..._
- [x] Log button should not be disabled during format/checksum operations
- [x] Reposition the multi toolbar buttons. LANG first, LOG last?
- [x] ~~Flat button color change on mouse over?~~ _Nah_
- [x] Remove positioning for main dialog from loc file
- [x] Split Partition Scheme and Target System
- [x] Link PS & TS to formatting
- [x] Sort out enabling/disabling of Start button
- [x] Sort out enabling/disabling of MBR/GPT/BIOS/UEFI
- [x] Sort out tooltips for Partition Scheme and Target System
- [x] Sort out DD vs ISO mode
- [x] FAT32 should not be allowed for Windows ISOs in BIOS mode
- [x] ~~3 levels of advanced mode (simple, normal, advanced)?~~ _Nah - Would only have worked in the old scheme anyway._
- [x] Resize all buttons so they have the same height (including Log)
- [x] Disable Start button if image option is selected and no image is loaded
- [x] ~~Do we want to move the `#` button to multi toolbar?~~ _Not for now. Will see if users complain..._
- [x] ~~Do we want to replace the `#` button with a toolbar `✓`?~~ _No. We tried it and it doesn't look that great_
- [x] Should we disable all controls while no image is selected? _Yes!_
- [x] Quick format gets re-enabled when hiding/showing advanced Drive Props
- [x] re-enable controls/previous image settings if new image fails to scan (as seen with `debian-live-9.1.0-i386-lxde.iso`)
- [x] Sort out the enabling/disabling of advanced options as well as boot selection dropdown
- [x] Don't re-display drive props in the log when switching boot type
- [x] Use `toupper` or whatever to convert our main buttons to uppercase. Might need an exception for _"Schließen"_ &rarr; _"SCHLIESSEN"_
- [x] `MSG_032 UNTRANSLATED` for Thai? WTF?!? _Was missing a `b "en-US"`_
- [x] Update French translation
- [x] Update all translations to and set `IDS_BOOT_SELECTION_TXT` to `MSG_278`
- [x] Update all translations so that it doesn't crash on MBR/GPT/BIOS/UEFI
- [x] Selecting Windows To Go should change FS to NTFS
- [x] Use GPT for UEFI by default on Windows BIOS + EFI images, instead of MBR for BIOS
- [x] Fix NTFS not being selected when picking a Windows ISO
- [x] Improve fallback for the up/down icons
- [x] And now we're getting a bunch of `MSG_000 UNTRANSLATED` during ISO extraction...
- [x] Add a tooltip for the Save button
- [x] Resize the friggin label button so it has a proper height
- [x] Tie hash computation to new button
- [x] Grey/hide hash button if not appropriate
- [x] ~~Populate Selected Image dropdown and keep history~~
- [x] Dammit: we don't have up and down icons for Advanced Options on Windows 7!
- [x] Set tooltips for Windows To Go and stuff
- [x] Tip/`?` about CSM appears when dropdown is not populated &mdash; avoid that.
- [x] Add a question mark tooltip at the end of the PS&TS dropdown that explains what UEFI-CSM is
- [x] Also add a question mark tooltip to explain what _non CSM_ means
- [x] _non CSM_ tooltip doesn't get set when selecting a Windows image
- [x] Remove all positioning on all dialogs (including `IDD_NOTIFICATION`)
- [x] Move `IDC_ABOUT` and `IDC_LOG` to `MSG_305/306` or introduce new tooltips
- [x] Remove `MSG_086/MSG_087`
- [x] `MSG_086`. Do we need it back?
- [x] Test and fix the UI for zoom factors other than 200%:
  - [x] No info icon for 16px
  - [x] Section headers and line width is too big for 100%
  - [x] Spacing between _Drive Properties_ &rarr; _Device_ and _Format Options_ &rarr; _Volume Label_ seems off
  - [x] Buttons are not the same height as dropdowns
  - [x] Status font is too big
  - [x] Items are too close to one another at 100% &mdash; we need to vertically space everything 1 or 2 em more at that DPI.
  - [x] Right margin is smaller than left margin at 100% &mdash; Whyyyyyyyyy?
  - [x] Advanced section toolbars eat on the checkboxes
  - [x] Triangle buttons are not aligned with the checkboxes
  - [x] Toolbar button spacing sucks at 100%
  - [x] If button height it smaller than ddh on additional dialogs, resize them
- [x] Windows 7 issues:
  - [x] One toolbar icon is missing
- [x] Go over the various `TODO`'s still in the code
- [x] Oh great, now we're seeing UI differences with the margins and toolbars between MinGW compiled code and MSVC compiled code. Really trying my patience here, Microsoft!
- [x] Coalesce all UI commits into a single one, then review that before committing to master
- [x] Carefully review all translation changes and update loc Changelog
- [x] Do we want to have a "No Device (Please connect a device)" message when we don't have anything in the device dropdown? _Nah, not worth it_
- [x] Produce an ALPHA, and notify translators about required translations

__Post Translator Alpha:__
- [x] Use MBR for BIOS + UEFI by default on Linux dual BIOS+UEFI images
- [x] We may want to prefer MBR as the partition scheme for non bootable
- [x] Sort other MBR vs GPT issues, including dual BIOS+EFI, and try to reselect current PS by default
- [x] Looks like we left a bunch of `->` over in the Malay translation. Fix that.
- [x] Status bar text is not properly vertically aligned at 150% and 200% &mdash; Should we make the status bar height bigger?
- [x] ~~Progress bar text overlay is not vertically centred~~ _Should (hopefully) be sorted with our custom control_
- [x] If you have FreeDOS selected and are showing the advanced format options, the _Quick Format_ option becomes disabled is you show the advanced drive properties... _Was due to Large FAT32 options inconsistency_
- [x] Windows To Go should remain selected if a user changes boot selection and back
- [x] Windows To Go/Image options are not preserved when switching languages
- [x] Improve fs reselection with `SetFileSystemAndClusterSize()`
- [x] Create our own 2-color icons for About/Settings/Language/Log/Save
- [x] ~~Grey-inverted buttons for SELECT, START, CANCEL~~ _Too much work_
- [x] Fix what happens when selecting an ISO with no UFD plugged, and then plugging a UFD
- [x] Partitioning as non-bootable/GPT/NTFS doesn't work
- [x] Button icons are __not__ centered on the multitoolbar
- [x] Displaying the advanced options forces a redraw of the whole screen (e.g. if you have UltraEdit open in the background). That isn't the case with 2.18. WTF is going on?!? _Seems to have to to with `InvalidateRect(hCtrl, NULL, TRUE);` in `ResizeMoveCtrl()`. Can we drop it?_
- [x] Create our own progress + text overlay bar, since Windows is useless
- [x] Add marquee mode to our progress bar
- [x] Do we care about the fact that we lose our Advanced toolbars at 100% DPI for the MSVC compiled version (but not the MinGW one) when advanced mode is enabled? _We do!_
- [x] ~~Use our custom progress bar to merge progress + download URL on new version dialog?~~ _Nah, not worth it for the 3.0 release..._
- [x] Progress bar doesn't behave properly when writing ISOHybrids in DD mode
- [x] `write error` message doesn't display the error message &rarr; _Looks like if you don't use `%lld` for large integers, the next variable pointer gets screwed up_
- [x] Extreme left side of progress bar gets erased when starting to zero a large drive (e.g. 1.5 TB HDD)
- [x] Can't seem to zero a drive after writing `archlinux-2018.01.01-x86_64.iso` in DD mode (`access denied`) unless we first reformat to FreeDOS... &rarr; _I think the issue has to do with having a handle to the actual Logical Drive before we can write sectors onto it. Because the partition is not recognized, we don't get an LD and therefore Windows shuts us down as soon as we reach it. Not sure what a good workaround would be at this stage..._
- [x] We also seem to get 'timeout while waiting for logical drive' after writing a DD image (e.g. Debian Live). Win10 1803 issue? &rarr; _Most likely a non-issue. If Windows doesn't recognize the partition, we of course won't be able to mount the LD, hence the timeout_
- [x] Also in Win10 1803, UFD is not being listed after DD image write in safe eject list. Probably linked to above... &rarr; _Yup, that one's a pure Windows issue_
- [x] Might need to look for FAT label to identify an UEFI:NTFS partitions as some distros (e.g. Debian Live) use the `0xef` partition code with their own small EFI boot partition
- [x] Can we do something about the margins not being taken into account when displaying the log window?
- [x] This _YubiKey Smart Card Minidriver_ automatic reinstallation by Windows is __really__ starting to bug me, and the workaround from [here](https://forum.yubico.com/viewtopic948b.html?p=9916) doesn't seem to work. Is there anything we can do to __force__ uninstall that crap before we call `signtool`? &rarr; `gpedit` driver installation seems to work now, at long last!
- [x] Sort out tabbing order, especially for the new toolbars and toolbar buttons
- [x] That extra tab to the disabled hash toolbar button is annoying...
- [x] Do we want to tab to the CSM help blurb? If so, how? &rarr; _Not for now... We'll look into it if we are asked to._
- [x] Is there any way to force tabbing __within__ the multitoolbar (instead of having to switch to using arrows instead of tab)? &rarr; _As per [Microsoft official documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/dn742395.aspx): "Once a menu bar has input focus, it is navigated independently of the remainder of the window, handling all arrow keys"_
- [x] Is there any way to silently label our toolbar buttons to help visually impaired people using a screen reader? &rarr; _Will look into it if asked..._
- [x] Do we need to care about making sure we don't disable a control that has focus as per [this](https://blogs.msdn.microsoft.com/oldnewthing/20040804-00/?p=38243)? &rarr; _I don't think so: As far as I can see, if someone uses keyboard to enable/disable stuff the control that triggered the disabling will never be disabled itself so this should be a non-issue._
- [x] Do we want to change the color of our `?` tooltip to match the toolbar icons? &rarr; _We do!_
- [x] We have to _unmirror_ some of the toolbar icons for RTL languages, notably `(✓)`. Maybe [this](https://docs.microsoft.com/en-us/globalization/localizability/mirroring-in-win32) can help?
- [x] I think our RTL tooltips are broken &mdash; Why the £$^& are they not right justified?
- [x] Anything we can do to improve RTL switching in UEFI/CSM messages...? &rarr; _I'll leave that as a job to __tomorrow me__! (i.e. moved to 3.1)_
- [x] `MSG_199/200` are no longer needed
- [x] ...and now the toolbar icons are not centered on Windows 7. Seriously Microsoft, WTF?!? &rarr; _Looks like the root issue is that centering is off at 1x scaling. Sigh..._
- [x] If you create a Windows 10 UFD as MBR/NTFS, once the creation is complete, partition scheme will switch to GPT...
- [x] We get seem to get persistent/unwanted `Show the log` tooltips while a formatting operation is in progress. e.g. Create a Windows 10 USB, open the log, close the log, move cursor back towards main dailog. Looks like this happens as soon as main dialog gets the log button highlighted an you move the mouse... Might work around this by setting focus to _Cancel_ once the format operation is started (which is what I thought we were already doing anyway?)
- [x] That note about _"clicking the white triangle"_ in the log is no longer accurate.
- [x] Chase translators about missing messages highlighted by `.loc` validation script
- [x] Try to get at least 80% of the translations updated

__Translation Update Status (34/38 = 89.5%):__
- [x] Arabic
- [x] Azerbaijani (Latin)
- [ ] Bulgarian
- [x] Chinese (Simplified)
- [x] Chinese (Traditional)
- [x] Croatian
- [x] Czech
- [x] Danish
- [x] Dutch
- [x] Finnish
- [x] French
- [x] German
- [x] Greek
- [x] Hebrew
- [x] Hungarian
- [x] Indonesian
- [x] Italian
- [x] Japanese
- [x] Korean
- [x] Latvian
- [x] Lithuanian
- [ ] Malay
- [x] Norwegian
- [x] Persian
- [x] Polish
- [x] Portuguese (Brazil)
- [x] Portuguese (Standard)
- [x] Romanian
- [x] Russian
- [ ] Serbian (Latin)
- [x] Slovak
- [x] Slovenian
- [x] Spanish
- [x] Swedish
- [x] Thai
- [x] Turkish
- [x] Ukrainian
- [ ] Vietnamese

**Comment by NSbuilder** [comment]
The next version is going to be 3.0 or 2.12?

**Comment by pbatard** [comment]
No idea. It'll be what it will be. Also, please avoid commenting on issues that are marked as __[INTERNAL]__. As their name indicate, they're just for internal notes, which can be ALTERED IN ANY DIRECTION AT ANY MOMENT and do __NOT__ mean that what you see written will actually ever be done.

**Comment by pbatard** [comment]
Also please see the [milestones](https://github.com/pbatard/rufus/milestones) if you are interested in the VERY VAGUE PLAN that I have for upcoming releases. This too can be altered drastically at any time, so please don't try to read too much into it.

**Comment by positr0nic** [comment]
I cant seem to find a way to make a bootable usb with persistence via rufus,  if thats is not supported it would be a fine addition.

**Comment by pbatard** [comment]
Please don't use issues flagged as [INTERNAL] to ask for features, especially as, if you had created a new issue, and followed the checklist, you would have found #691.

**Comment by positr0nic** [comment]
Apologies, its not clear anywhere that this is not a thing todo. I simply saw a checklist and thought it was relevant.
